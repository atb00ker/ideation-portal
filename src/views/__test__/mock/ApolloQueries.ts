import { MockedResponse } from '@apollo/client/testing';
import { TopicCategory } from '../../enums/TopicCategory';
import { TopicDepartment } from '../../enums/TopicDepartment';
import { TopicStatus } from '../../enums/TopicStatus';
import { GET_EXCEL_QUERY } from '../../graphql/GetExcel';
import { GET_REPORT_QUERY } from '../../graphql/GetReport';
import { GET_TOPIC_COLLECTION } from '../../graphql/GetTopicsCollection';
import { IGetTopicCollectionInput } from '../../interfaces/IGetTopicCollectionInput';

export const MOCK_GET_TOPIC_COLLECTION: MockedResponse = {
  request: {
    query: GET_TOPIC_COLLECTION,
    variables: {
      limit: parseInt(process.env.HOME_RECORDS_PAGE_SIZE || '5'),
      offset: 0,
      searchFilter: '',
      categoryIdList: Object.keys(TopicCategory).map(key => TopicCategory[key].id),
      departmentIdList: Object.keys(TopicDepartment).map(key => TopicDepartment[key].id),
      statusIdList: Object.keys(TopicStatus).map(key => TopicStatus[key].id),
    } as IGetTopicCollectionInput,
  },
  result: {
    data: {
      topics: [
        {
          id: 'testId',
          title: 'testTitle',
          short_description: 'testDescription',
          status: 0,
          category: 0,
          department: 0,
          topics_users_likes_associations_aggregate: {
            aggregate: {
              count: 10,
            },
          },
        },
      ],
      topics_aggregate: {
        aggregate: {
          count: 10,
        },
      },
    },
  },
};

export const MOCK_GET_EXCEL_QUERY: MockedResponse = {
  request: {
    query: GET_EXCEL_QUERY,
  },
  result: {
    data: {
      GetExcel: {
        excel_link: 'reports.csv',
      },
    },
  },
};

export const MOCK_GET_REPORT: MockedResponse = {
  request: {
    query: GET_REPORT_QUERY,
  },
  result: {
    data: {
      reports_by_pk: {
        site_visitors: 100,
      },
      topics_count_by_category: [
        {
          category: 0,
          count: 1,
        },
      ],
      topics_count_by_status: [
        {
          status: 0,
          count: 1,
        },
      ],
      topics_count_by_department: [
        {
          department: 0,
          count: 1,
        },
      ],
      topics_created_by_month: [
        {
          month: 'Jan',
          topics_created: 10,
        },
      ],
    },
  },
};
