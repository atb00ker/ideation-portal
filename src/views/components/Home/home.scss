/**************************
* Search Bar Section
**************************/

#search-icon {
  transform: translateX(8px);
  position: absolute;
  height: 34px;
  color: gray;
}

#search-input {
  width: 100%;
  max-width: 960px;
  padding-left: 23px;
}
