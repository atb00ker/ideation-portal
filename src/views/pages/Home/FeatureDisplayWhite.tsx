import React from 'react';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';
import Image from 'react-bootstrap/Image';
import Container from 'react-bootstrap/Container';
import { FeatureDisplayType } from '../../types/react-component-input';

const FeatureDisplayWhite: React.FC<FeatureDisplayType> = ({ image, title, description, imageWidth }) => {
  return (
    <Row className='cover-screen'>
      <Col sm='12'>
        <Container className='h-100 max-width-960'>
          <Row className='h-100'>
            <Col xs='12' md='4'>
              <div className='cover-screen-flex'>
                <Row>
                  <Col xs='12'>
                    <h2 className='text-primary'>{title}</h2>
                  </Col>
                  <Col xs='12'>{description}</Col>
                </Row>
              </div>
            </Col>
            <Col
              className='cover-screen-flex'
              xs={{ order: 'first', span: '12' }}
              md={{ order: 'last', span: '8' }}>
              <Image
                className='d-none d-md-block cover-screen-flex'
                src={image}
                alt={title}
                width={imageWidth}
              />
              <Image className='d-md-none cover-screen-flex' src={image} alt={title} width={'100%'} />{' '}
            </Col>
          </Row>
        </Container>
      </Col>
    </Row>
  );
};

export { FeatureDisplayWhite };
