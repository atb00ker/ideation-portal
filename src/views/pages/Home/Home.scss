.cover-screen {
  height: calc(100vh - 60px);
}

.bg-off-white {
  background-color: whitesmoke;
  border-color: whitesmoke;
  border-radius: 25% 2% 25% 2%;
  border-style: solid;
}

.cover-screen-flex {
  display: flex;
  justify-content: center;
  align-items: center;
}

@media (min-width: 768px) {
  .cover-screen-flex {
    height: 100%;
  }
}

@media (max-width: 768px) {
  .text-center-md {
    text-align: center;
  }

  .bg-off-white {
    border-radius: 10% 1% 10% 1%;
  }
}
