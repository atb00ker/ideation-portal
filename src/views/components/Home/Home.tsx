import React from 'react';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Card from 'react-bootstrap/Card';
import Stepper from '@material-ui/core/Stepper';
import Step from '@material-ui/core/Step';
import StepLabel from '@material-ui/core/StepLabel';
import './home.scss';

const Home = () => {
  return (
    <Container>
      <Row>
        <Col xs={12}>
          <div className="mx-auto mt-3 input-position text-center">
            <svg id="search-icon" xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" className="bi bi-search" viewBox="0 0 16 16">
              <path d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z" />
            </svg>
            <input id="search-input" type="search" placeholder="Search" />
          </div>
        </Col>
      </Row>
      <Row>
        <Col xs={12}>
          <Card className="idea-card mx-auto mt-4">
            <Card.Body>
              <Card.Title>Ideation Portal</Card.Title>
              <Card.Subtitle className="mb-2 text-muted">Mindset - Product Engineering - Ajay Tripathi</Card.Subtitle>
              <Card.Text>
                Some quick example text to build on the card title and make up the bulk of
                the card's content. Some quick example text to build on the card title and make up the
                bulk of the card's content.
              </Card.Text>
            </Card.Body>
            <div className="card-options">
              <Stepper className="card-stepper-container" activeStep={2}>
                <Step key={1}>
                  <StepLabel>Proposed</StepLabel>
                </Step>
                <Step key={2}>
                  <StepLabel>Brainstorming</StepLabel>
                </Step>
                <Step key={3}>
                  <StepLabel>Creating</StepLabel>
                </Step>
                <Step key={4}>
                  <StepLabel>Feedback</StepLabel>
                </Step>
                <Step key={4}>
                  <StepLabel>Completed</StepLabel>
                </Step>
              </Stepper>
              <Card.Link className="rounded-0 card-btn btn btn-primary" href="#">Read More</Card.Link>
            </div>
          </Card>
        </Col>
      </Row>
    </Container>
  );
};

export default Home;
