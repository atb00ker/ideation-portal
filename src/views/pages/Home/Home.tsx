import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Button from 'react-bootstrap/Button';
import Pagination from 'react-bootstrap/Pagination'
import { ErrorBoundary } from 'react-error-boundary';

import { ErrorFallback } from '../../components/ContentState/ErrorFallback';
import TopicsCollection from '../../components/TopicsCollection/TopicsCollection';
import GetTopicsCollection from '../../components/GraphQL/GetTopicsCollection';
import SectionLoader from '../../components/ContentState/SectionLoader';
import ServerRequestError from '../../components/ContentState/ServerRequestError';
import NoRecordsFound from '../../components/ContentState/NoRecordsFound';
import GetTopicsCount from '../../components/GraphQL/GetTopicsCount';
import SearchSvg from '../../assets/icons/bootstrap-search.svg';
import './home.scss';

const Home = () => {
  let pageSize = 5;
  const [pageActive, setPageActive] = useState(0);
  const { loading: countLoading, data: countData, error: countError } = GetTopicsCount();
  const { loading: topicsLoading, data: topicsData,
          error: topicsError } = GetTopicsCollection(pageSize, pageSize*pageActive);

  const AddPageSection = (
    <Row className="mt-4">
      <Col xs={12}>
        <Row className="m-auto max-width-960">
          <Col className="ps-0" sm={10} md={11}>
            Do you have an Idea to share?
            A challenge your team is facing?
            We want to hear about it! Please click the 'add' button to share.
          </Col>
          <Col className="pe-0" sm={2} md={1}>
            <Link to={'/topic/'}>
            <Button className="float-end" variant="primary">Add</Button>
            </Link>
          </Col>
        </Row>
      </Col>
    </Row>
  );

  if (countLoading || topicsLoading)
    return <SectionLoader height="500px" width="100%" />

  if (countError || topicsError) {
    console.error(countError, topicsError);
    return <ServerRequestError height="500px" imgHeight="250px" width="100%" />
  }

  if (!topicsData.topics.length) {
    return (
      <>
        {AddPageSection}
        <NoRecordsFound height="500px" imgHeight="250px" width="100%" />
      </>
    )
  }

  const changePage = (pageNumber: number) => setPageActive(pageNumber);
  const pagesCount = Math.ceil(countData.topics_aggregate.aggregate.count / pageSize);

  return (
    <Container>
      {AddPageSection}
      <Row>
        <Col xs={12}>
          <ErrorBoundary FallbackComponent={ErrorFallback}>
            <div className="mx-auto mt-2 input-position text-center">
              <svg id="search-icon" xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" className="bi bi-search" viewBox="0 0 16 16">
                <path d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z" />
              </svg>
              <input id="search-input" className="max-width-960" type="search" placeholder="Search" />
            </div>
          </ErrorBoundary>
        </Col>
      </Row>
      <Row className="max-width-960 mx-auto">
        <Col xs={12} className="mt-2 mb-5 p-0">
          <ErrorBoundary FallbackComponent={ErrorFallback}>
            <TopicsCollection topics={topicsData.topics} />
            {pagesCount > 1 && <Pagination className="mb-4 mt-4 float-end" size="sm">
              <Pagination.Prev />
              {Array.from(Array(pagesCount), (_, index) => {
                return <Pagination.Item onClick={() => changePage(index)} key={index}
                                        active={index==pageActive}>
                  {index + 1}
                </Pagination.Item>
              })}
              <Pagination.Next />
            </Pagination>}
          </ErrorBoundary>
        </Col>
      </Row>
    </Container>
  );
};

export default Home;
