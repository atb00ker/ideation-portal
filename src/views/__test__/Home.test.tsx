/**
 * @jest-environment jsdom
 */

import { MockedProvider as ApolloMockedProvider } from '@apollo/react-testing';
import '@testing-library/jest-dom';
import 'regenerator-runtime/runtime';
import React from 'react';
import ReactDOM from 'react-dom';
import { render, cleanup, act, waitFor, RenderResult } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { HashRouter as Router } from 'react-router-dom';
import Home from '../pages/Home/Home';
import { MOCK_GET_TOPIC_COLLECTION } from './mock/ApolloQueries';

const _homeComponent = (
  <ApolloMockedProvider addTypename={false} mocks={[MOCK_GET_TOPIC_COLLECTION]}>
    <Router>
      <Home></Home>
    </Router>
  </ApolloMockedProvider>
);

describe('Home page tests', () => {
  // beforeEach(() => {});
  afterEach(() => cleanup);

  it('renders without crashing', () => {
    const root = document.createElement('div');
    ReactDOM.render(_homeComponent, root);
  });

  it('test search and filter section exists and content section exists', async () => {
    let page: RenderResult | any;
    await act(async () => {
      page = await render(_homeComponent);
    });
    expect(page.getByText('All Statuses')).toBeInTheDocument();
    expect(page.getByText('All Departments')).toBeInTheDocument();
    expect(page.getByText('All Categories')).toBeInTheDocument();
    expect(page.getByTestId('home-top-message')).toBeInTheDocument();
    expect(page.getByTestId('home-search-box')).toBeInTheDocument();
    await waitFor(() => expect(page.getByText('testTitle')).toBeInTheDocument());
  });

  it('home snapshot structure', () => {
    const snapTree = renderer.create(_homeComponent);
    expect(snapTree.toJSON()).toMatchSnapshot();
  });
});
