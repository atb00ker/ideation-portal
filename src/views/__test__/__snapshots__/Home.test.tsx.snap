// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page tests home snapshot match 1`] = `
<div
  className="container"
>
  <div
    className="mt-4 row"
  >
    <div
      className="col-12"
    >
      <div
        className="m-auto max-width-960 row"
      >
        <div
          className="ps-0 col-md-11 col-sm-10"
          data-testid="home-top-message"
        >
          Do you have an Idea to share? A challenge your team is facing? We want to hear about it! Please click the 'add' button to share.
        </div>
        <div
          className="pe-0 col-md-1 col-sm-2"
        >
          <a
            href="#/topic/"
            onClick={[Function]}
          >
            <button
              className="float-end btn btn-primary"
              disabled={false}
              type="button"
            >
              Add
            </button>
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    className="max-width-960 mx-auto row"
  >
    <div
      className="p-0 col-12"
    >
      <div
        className="mt-2"
      >
        <form
          className=""
          onSubmit={[Function]}
        >
          <div
            className="mb-3 d-flex"
          >
            <input
              className="form-control"
              data-testid="home-search-box"
              id="search-input"
              placeholder="Search"
              type="search"
            />
            <button
              aria-label="Click to Search"
              className="btn btn-primary"
              disabled={false}
              id="search-icon-btn"
              title="Click to Search"
              type="submit"
            >
              <svg
                className="bi bi-search"
                fill="currentColor"
                id="search-icon"
                viewBox="0 -3 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"
                />
              </svg>
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
  <div
    className="max-width-960 mx-auto row"
  >
    <div
      className="mb-3 p-0 me-1 col-3"
    >
      <form
        className=""
      >
        <div
          className="mb-3"
        >
          <select
            className="form-select"
            defaultValue="All Categories"
            id="AllCategories"
            onChange={[Function]}
          >
            <option
              value={
                Array [
                  0,
                  1,
                ]
              }
            >
              All Categories
            </option>
            <option
              value={0}
            >
              Mindset
            </option>
            <option
              value={1}
            >
              Customer Delight
            </option>
          </select>
        </div>
      </form>
    </div>
    <div
      className="mb-3 p-0 me-1 col-3"
    >
      <form
        className=""
      >
        <div
          className="mb-3"
        >
          <select
            className="form-select"
            defaultValue="All Departments"
            id="AllDepartments"
            onChange={[Function]}
          >
            <option
              value={
                Array [
                  0,
                  1,
                ]
              }
            >
              All Departments
            </option>
            <option
              value={0}
            >
              Product Engineering
            </option>
            <option
              value={1}
            >
              Product Management
            </option>
          </select>
        </div>
      </form>
    </div>
    <div
      className="mb-3 p-0 me-1 col-3"
    >
      <form
        className=""
      >
        <div
          className="mb-3"
        >
          <select
            className="form-select"
            defaultValue="All Statuses"
            id="AllStatuses"
            onChange={[Function]}
          >
            <option
              value={
                Array [
                  0,
                  1,
                  2,
                  3,
                  4,
                ]
              }
            >
              All Statuses
            </option>
            <option
              value={0}
            >
              Proposed
            </option>
            <option
              value={1}
            >
              Brainstorming
            </option>
            <option
              value={2}
            >
              Creating
            </option>
            <option
              value={3}
            >
              Feedback
            </option>
            <option
              value={4}
            >
              Completed
            </option>
          </select>
        </div>
      </form>
    </div>
  </div>
  <div
    className="position-relative"
    style={
      Object {
        "height": "500px",
        "width": "100%",
      }
    }
  >
    <div
      className="round-loader-center position-absolute spinner-border"
      role="status"
    >
      <span
        className="visually-hidden"
        data-testid="section-loading"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;

exports[`Home page tests home snapshot structure 1`] = `
<div
  className="container"
>
  <div
    className="mt-4 row"
  >
    <div
      className="col-12"
    >
      <div
        className="m-auto max-width-960 row"
      >
        <div
          className="ps-0 col-md-11 col-sm-10"
          data-testid="home-top-message"
        >
          Do you have an Idea to share? A challenge your team is facing? We want to hear about it! Please click the 'add' button to share.
        </div>
        <div
          className="pe-0 col-md-1 col-sm-2"
        >
          <a
            href="#/topic/"
            onClick={[Function]}
          >
            <button
              className="float-end btn btn-primary"
              disabled={false}
              type="button"
            >
              Add
            </button>
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    className="max-width-960 mx-auto row"
  >
    <div
      className="p-0 col-12"
    >
      <div
        className="mt-2"
      >
        <form
          className=""
          onSubmit={[Function]}
        >
          <div
            className="mb-3 d-flex"
          >
            <input
              className="form-control"
              data-testid="home-search-box"
              id="search-input"
              placeholder="Search"
              type="search"
            />
            <button
              aria-label="Click to Search"
              className="btn btn-primary"
              disabled={false}
              id="search-icon-btn"
              title="Click to Search"
              type="submit"
            >
              <svg
                className="bi bi-search"
                fill="currentColor"
                id="search-icon"
                viewBox="0 -3 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"
                />
              </svg>
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
  <div
    className="max-width-960 mx-auto row"
  >
    <div
      className="mb-3 p-0 me-1 col-3"
    >
      <form
        className=""
      >
        <div
          className="mb-3"
        >
          <select
            className="form-select"
            defaultValue="All Categories"
            id="AllCategories"
            onChange={[Function]}
          >
            <option
              value={
                Array [
                  0,
                  1,
                ]
              }
            >
              All Categories
            </option>
            <option
              value={0}
            >
              Mindset
            </option>
            <option
              value={1}
            >
              Customer Delight
            </option>
          </select>
        </div>
      </form>
    </div>
    <div
      className="mb-3 p-0 me-1 col-3"
    >
      <form
        className=""
      >
        <div
          className="mb-3"
        >
          <select
            className="form-select"
            defaultValue="All Departments"
            id="AllDepartments"
            onChange={[Function]}
          >
            <option
              value={
                Array [
                  0,
                  1,
                ]
              }
            >
              All Departments
            </option>
            <option
              value={0}
            >
              Product Engineering
            </option>
            <option
              value={1}
            >
              Product Management
            </option>
          </select>
        </div>
      </form>
    </div>
    <div
      className="mb-3 p-0 me-1 col-3"
    >
      <form
        className=""
      >
        <div
          className="mb-3"
        >
          <select
            className="form-select"
            defaultValue="All Statuses"
            id="AllStatuses"
            onChange={[Function]}
          >
            <option
              value={
                Array [
                  0,
                  1,
                  2,
                  3,
                  4,
                ]
              }
            >
              All Statuses
            </option>
            <option
              value={0}
            >
              Proposed
            </option>
            <option
              value={1}
            >
              Brainstorming
            </option>
            <option
              value={2}
            >
              Creating
            </option>
            <option
              value={3}
            >
              Feedback
            </option>
            <option
              value={4}
            >
              Completed
            </option>
          </select>
        </div>
      </form>
    </div>
  </div>
  <div
    className="position-relative"
    style={
      Object {
        "height": "500px",
        "width": "100%",
      }
    }
  >
    <div
      className="round-loader-center position-absolute spinner-border"
      role="status"
    >
      <span
        className="visually-hidden"
        data-testid="section-loading"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;
